Remove the boost/algorithm/string/predicate.hpp dependency

l2a5b1 opened the pull request:
This pull request removes the `boost/algorithm/string/predicate.hpp` dependency from the project.

To replace the the `predicate.hpp` dependency from the project the function calls to `boost::algorithm::starts_with` and `boost::algorithm::ends_with` have been replaced with respectively C++11's `std::basic_string::front` and `std::basic_string::back` function calls.

Refactors that were not required, but have been done anyways:

- The Boost function `all` was implicitly made available via the `predicate.hpp` header. Instead of including the appropriate header, function calls to `all` have been replaced with function calls to `std::all_of`.

- The  `boost::algorithm::is_digit` predicate has been replaced with a custom `IsDigit` function that is locale independent and ASCII deterministic.

MarcoFalke commented:
utACK 6285655d9592bf684f2ddc69871456fb3628b8a6

Empact commented:
utACK 6285655

fanquake commented:
Concept ACK

After this change there are only a few other uses of `starts_with` and `ends_with` left in the code base, all in `core_read.cpp` and `src/wallet/rpcdump.cpp`. @251Labs Did you want to update this PR to remove all the remaining instances in one go? 

If you do, please also remove `<boost/algorithm/string/predicate.hpp>` from [lint-includes.sh](https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/test/lint/lint-includes.sh#L53)

##### starts_with
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/core_read.cpp#L57
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/core_read.cpp#L63
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/core_read.cpp#L69
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/wallet/rpcdump.cpp#L588
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/wallet/rpcdump.cpp#L594

##### ends_with
https://github.com/bitcoin/bitcoin/blob/8803c9132a78d8182bd828a29f7051fc7688f934/src/core_read.cpp#L69

Empact commented:
Could follow @fanquake's suggestion for just the single-character checks.

251Labs commented:
Thanks @fanquake and @empact. I kept the PR as small as possible for ease of review, but I am happy to have a go at the remaining instances.

Empact commented:
@251Labs dealing with all identical cases in a single PR can help minimize review effort to avoid splitting consideration and discussion over multiple PRs.

251Labs commented:
Thanks for the clarification @Empact. I have updated the PR and changed the title and description accordingly.

Empact commented on src/utilstrencodings.cpp:
```diff
@@ -425,6 +425,11 @@ int atoi(const std::string& str)
     return atoi(str.c_str());
 }
 
+bool IsDigit(char c)
```
nit: `constexpr`

251Labs replied:
Nice one @empact, addressed in 2a02b4a.

Empact commented:
There are 5 places in `utilstrencoding.cpp` where `IsDigit` could be applied. Could do that here or in a follow-up but here makes sense to me.

Note the other possible uses are in included projects, e..g leveldb, tinyformat, which shouldn't be changed.

251Labs commented:
Thanks again @Empact , I have addressed your feedback in 2a02b4a.

Empact commented:
utACK 146793c

251Labs commented:
Thanks @empact, I am happy to squash this. 

Just wanted to let you know that 2a02b4a is an atomic commit on its own. Strictly speaking the work in 2a02b4a was not required to drop the predicate dependency in 146793c. Maybe a squashed commit would be confusing from that POV (particularly in a limited blame- or history view).

Either way, I am happy to squash this. Please let me know what is preferred and I will take it from there :)

Empact commented:
Could go either way, but IMO neither is so complicated that they don't bear combination, and the latter motivates the former.

sipa commented:
utACK 146793caa9403310ae44d1f7fa54dc077d4b2d39

MarcoFalke commented on src/core_read.cpp:
```diff
@@ -53,20 +54,20 @@ CScript ParseScript(const std::string& s)
             result << n;
         }
-        else if (boost::algorithm::starts_with(*w, "0x") && (w->begin()+2 != w->end()) && IsHex(std::string(w->begin()+2, w->end())))
+        else if (w->substr(0,2) == "0x" && (w->begin()+2 < w->end()) && IsHex(std::string(w->begin()+2, w->end())))
```
unrelated nit: could use `std::string::size > 2` here

MarcoFalke commented:
utACK 146793caa9403310ae44d1f7fa54dc077d4b2d39

l2a5b1 committed:
Removes Boost predicate.hpp dependency

This is a squashed commit that squashes the following commits:

This commit removes the `boost/algorithm/string/predicate.hpp` dependenc
from the project by replacing the function calls to `boost::algorithm::starts_with`
`boost::algorithm::ends_with` and `all` with respectively C++11'
`std::basic_string::front`, `std::basic_string::back`, `std::all_of` function calls

This commit replaces `boost::algorithm::is_digit` with  a locale independent isdigi
function, because the use of the standard library's `isdigit` and `std::isdigit
functions is discoraged in the developer notes

251Labs commented:
Thanks for the reviews! Not sure if it is appreciated if nits are addressed following utACKs, but e3245f2 addresses @MarcoFalke's nit https://github.com/bitcoin/bitcoin/pull/13656#discussion_r204157110; squashes individual commits per @Empact's feedback; and is rebased on master (0a34593).

Empact commented on src/core_read.cpp:
```diff
@@ -54,20 +55,20 @@ CScript ParseScript(const std::string& s)
         {
             // Empty string, ignore. (boost::split given '' will return one word)
```
unrelated nit: could `continue` here to make this explicit

Empact commented:
utACK e3245f2

MarcoFalke commented:
re-utACK e3245f2e7b6f98cda38a3806da854f7d513fec2f
